Scheduler panics (nil pointer) in preempt when a scenario has no tasks to allocate

### Summary
The scheduler panics with a nil-pointer dereference during the `preempt` action's scenario building. With ≥2 replicas this manifests as a leader-election crash-loop/flap (the leader dies mid-cycle, a standby takes over, repeat) and intermittent scheduling stalls. Running pods are unaffected (control-plane only).

### Panic
```
panic: runtime error: invalid memory address or nil pointer dereference
  .../accumulated_scenario_filters/idle_gpus.NewIdleGpusFilter(0x0, ...)   idle_gpus.go:57
  solvers.NewPodAccumulatedScenarioBuilder(...)                            pod_scenario_builder.go:65
  solvers.(*JobSolver).solvePartialJob(...)                                job_solver.go
  solvers.(*JobSolver).Solve(...)
  preempt.attemptToPreemptForPreemptor(...)                                preempt/preempt.go
  scheduler.(*Scheduler).runOnce(...)
```

### Root cause
In `pod_scenario_builder.go`, `scenario` is only assigned when the (partial) job has tasks to allocate:
```go
var scenario *solverscenario.ByNodeScenario = nil
tasksToAllocate := podgroup_info.GetTasksToAllocate(pendingJob, ...)
if len(tasksToAllocate) != 0 {
    scenario = solverscenario.NewByNodeScenario(...)
}
...
idleGpusScenarioFilter := idle_gpus_filter.NewIdleGpusFilter(scenario, ...) // may be nil
```
`NewIdleGpusFilter` then dereferences it (`len(scenario.PendingTasks())`). Its sibling filters (`NewNodeAffinitiesFilter`, `NewTopologyAwareIdleGpusFilter`) already nil-check the scenario; `NewIdleGpusFilter` does not.

### When it fires
The nil scenario arises in the preempt solver's scenario search (`solvePartialJob`) for a preemptor whose partial job resolves to **zero allocatable tasks** — a rare, transient state rather than a deterministic function of a single cluster state:
- It needs a brief alignment (which of an elastic job's pods are allocated vs pending, available victims, node fragmentation), so it surfaces only intermittently even when candidate preemptors are present.
- Within a cycle, Go's randomized map-iteration order varies which preemptors/victims/scenarios are explored, so near-identical state can take different search paths.

As a result it is **not reproducible from a point-in-time snapshot**: replaying a snapshot captured during an active crash window — with the same binary version — through `snapshot-tool` (which runs the real actions) did **not** reproduce the panic across 40 runs. A snapshot freezes one instant and cannot capture the cross-cycle evolution that produces the trigger.

### Impact
Leader-elected scheduler crash-loops/flaps; intermittent scheduling gaps during failover; a preempt cycle that hits this aborts before committing. Running workloads unaffected.

### Affected versions
Observed on `v0.14.2`; the same nil-able `scenario` → `NewIdleGpusFilter` path is present on `main`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 - Improved solver performance in some large reclaim scenarios [#1627](https://github.com/kai-scheduler/KAI-Scheduler/pull/1627) [itsomri](https://github.com/itsomri)
 - Grove grouper now sets `minSubGroup` (equal to the number of child SubGroups) instead of `minMember=0` on parent SubGroups generated from `topologyConstraintGroupConfigs` [#1639](https://github.com/kai-scheduler/KAI-Scheduler/issues/1639) [davidLif](https://github.com/davidLif)
 - Fixed Helm chart not wiring `podgrouper.queueLabelKey` into `spec.global.queueLabelKey` on the Config CR, so custom queue label keys were ignored at install time [#1655](https://github.com/kai-scheduler/KAI-Scheduler/pull/1655) [dttung2905](https://github.com/dttung2905)
+- Fixed scheduler nil-pointer panic in the preempt scenario builder when a (partial) job has no tasks to allocate (`NewIdleGpusFilter` dereferenced a nil scenario); added the missing nil-guard matching the sibling filters [#1664](https://github.com/kai-scheduler/KAI-Scheduler/issues/1664) [sam-huang1223](https://github.com/sam-huang1223)
 
 ## [v0.15.0] - 2026-05-20
 

--- a/pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/idle_gpus.go
+++ b/pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/idle_gpus.go
@@ -53,6 +53,9 @@ type AccumulatedIdleGpus struct {
 
 func NewIdleGpusFilter(
 	scenario *scenario.ByNodeScenario, nodeInfosMap map[string]*node_info.NodeInfo) *AccumulatedIdleGpus {
+	if scenario == nil {
+		return nil
+	}
 	idleGpusMap, relevantNodesSorted := createGpuMap(nodeInfosMap, len(scenario.PendingTasks()))
 
 	filter := &AccumulatedIdleGpus{

--- a/pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/idle_gpus_test.go
+++ b/pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/idle_gpus_test.go
@@ -1378,3 +1378,9 @@ func TestAccumulatedIdleGpus_Filter(t *testing.T) {
 		})
 	}
 }
+
+func Test_NewIdleGpusFilter_NilScenario(t *testing.T) {
+	if filter := NewIdleGpusFilter(nil, map[string]*node_info.NodeInfo{}); filter != nil {
+		t.Fatalf("expected nil filter for nil scenario, got %#v", filter)
+	}
+}

--- a/pkg/scheduler/actions/common/solvers/pod_scenario_builder.go
+++ b/pkg/scheduler/actions/common/solvers/pod_scenario_builder.go
@@ -115,6 +115,9 @@ func (asb *PodAccumulatedScenarioBuilder) GetNextScenario() *solverscenario.ByNo
 // outer state. advanceFirst controls whether the first pass starts by popping a
 // victim or by evaluating the current state as-is.
 func (asb *PodAccumulatedScenarioBuilder) iterate(advanceFirst bool) *solverscenario.ByNodeScenario {
+	if asb.lastScenario == nil {
+		return nil
+	}
 	needAdvance := advanceFirst
 	for {
 		if sub := asb.nextFromSubEmitter(); sub != nil {

--- a/pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go
+++ b/pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go
@@ -284,6 +284,24 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
 			Expect(numberOfGeneratedScenarios).To(Equal(len(potentialVictimsPerScenario)))
 		})
 	})
+
+	Context("preemptor with no tasks to allocate - nil scenario", func() {
+		BeforeEach(func() {
+			ssn, _ = initializeSession(0, 0)
+			submitQueue := createQueue("team-a")
+			ssn.ClusterInfo.Queues[submitQueue.UID] = submitQueue
+			// Running tasks are not allocatable, so the job has no tasks to allocate.
+			reclaimerJob, _ = createJobWithTasks(1, 1, "team-a", v1.PodRunning, []v1.ResourceRequirements{})
+			recordedVictimsJobs := []*podgroup_info.PodGroupInfo{}
+			victimsQueue := utils.GetVictimsQueue(ssn, nil)
+
+			scenarioBuilder = NewPodAccumulatedScenarioBuilder(ssn, reclaimerJob, recordedVictimsJobs, victimsQueue, ssn.ClusterInfo.Nodes)
+		})
+		It("does not panic and yields no scenario", func() {
+			Expect(scenarioBuilder.GetValidScenario()).To(BeNil())
+			Expect(scenarioBuilder.GetNextScenario()).To(BeNil())
+		})
+	})
 })
 
 func initializeSession(jobsCount, tasksPerJob int) (*framework.Session, []*pod_info.PodInfo) {

